### PR TITLE
Bugfix (macOS): Remove PYTHONEXECUTABLE from `env` for conda installs

### LINKED ISSUE #5621: Bug: conda install of plugin via Plugin Manager doesn't work when run from terminal (macOS only?)
## 🐛 Bug

I originally noted this in #5198 
Using the `conda` source to install a plugin (pure conda env, everything from conda-forge) doesn't work and shows an error in the Status:
```
Starting '/Users/piotrsobolewski/Dev/miniforge3/bin/mamba' with args ['install', '-y', '--prefix', '/Users/piotrsobolewski/Dev/miniforge3/envs/napari-main', '--override-channels', '-c', 'conda-forge', 'napari-animation==0.0.5']
Traceback (most recent call last):
  File "/Users/piotrsobolewski/Dev/miniforge3/bin/mamba", line 7, in <module>

    from mamba.mamba import main
ModuleNotFoundError: No module named 'mamba'

Task finished with exit code 1 with status 0.
```
Note: I have mamba installed (mamba 1.3.0, conda 22.9.0).
 
## To Reproduce

1. Make a conda env with the latest main nightly: `mamba create -n napari-main -c "napari/label/nightly" -c conda-forge napari python=3.10`
2. run napari from the terminal
3. Open the Plugin Manager, scroll down to `napari-animation`
4. Open the Installation Info menu and change source from PyPi to conda
5. Click install, open Show Status

Note: this issue only occurs when running napari direct from the terminal and is almost certainly macOS-only.

## Expected behavior

The plugin should install, if the option to change to `conda` is available.

## Environment

```
napari: 0.5.0a2.dev42+g9e911040
Platform: macOS-13.2.1-arm64-arm-64bit
System: MacOS 13.2.1
Python: 3.10.9 | packaged by conda-forge | (main, Feb 2 2023, 20:26:08) [Clang 14.0.6 ]
Qt: 5.15.6
PyQt5: 5.15.7
NumPy: 1.24.2
SciPy: 1.10.1
Dask: 2023.3.1
VisPy: 0.12.1
magicgui: 0.7.2
superqt: unknown
in-n-out: 0.1.7
app-model: 0.1.2
npe2: 0.6.2

OpenGL:
- GL version: 2.1 Metal - 83
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1680x1050, scale 2.0

Settings path:
- /Users/piotrsobolewski/Library/Application Support/napari/napari-main_e240b8daeed1098c91bdc4f2a9ad585200c6d954/settings.yaml
```

## Additional context

`mamba` is installed in `base`. if I open python REPL, in base, then `from mamba.mamba import main` works. In my napari env, it doesn't, with the same error as above.
Also, a simple test of QProcess works:
```
from qtpy.QtCore import QProcess

def test_mamba_qprocess():
    process = QProcess()
    process.start('mamba', ['--version'])
    process.waitForFinished()

    if process.exitStatus() == QProcess.NormalExit:
        output = process.readAllStandardOutput().data().decode().strip()
        if output.startswith('mamba '):
            return output.split()[1]
    return None
```

### PR DESCRIPTION
**Pull request by psobolewskiPhD**
# Description

In https://github.com/napari/napari/pull/5531 PYTHONEXECUTABLE is set to `sys.executable` on macOS to fix an issue with venv/pyenv and launching napari.
It seems to have a side effect that makes conda plugin installations using the new Plugin Manager not work when napari is launched from the Terminal on macOS, probably because `sys.executable`, is in the env that napari was run from. But that env doesn't have conda/mamba, only `base` does.

So this change removes PYTHONEXECUTABLE from the QProcessEnvironment within the `CondaInstallerTool` class.

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
Similar issue over at spyder: https://github.com/spyder-ide/spyder/issues/20599
PR that added the PYTHONEXECUTABLE: #5531 cc: @aganders3 
Closes #5621 5621 

# How has this been tested?
tested locally on my mac

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by psobolewskiPhD** [comment]
So `env.remove()` doesn't seem to complain if the key doesn't exist, but if my testing is correct and this is only set on macOS in the first place (and only when launched from command line), then  maybe this should be behind `if os.name == "darwin":` like the `nt` stuff further down?

File: napari/_qt/dialogs/qt_package_installer.py [hunk 1]
```diff
@@ -201,6 +201,9 @@ def environment(
     ) -> QProcessEnvironment:
         if env is None:
             env = QProcessEnvironment.systemEnvironment()
+            # Fix for macOS when napari launched from terminal
+            # related to https://github.com/napari/napari/pull/5531
+            env.remove("PYTHONEXECUTABLE")
```
Comment (jni) [inline napari/_qt/dialogs/qt_package_installer.py]: Is env a reference or a copy? I think it should only be removed if it's a copy. I guess you can test this with:

```
env = QProcessEnvironment.systemEnvironment()
env.remove('PYTHONEXECUTABLE')
'PYTHONEXECUTABLE' in QProcessEnvironment.systemEnvironment()
```

If it's False, then I suggest:

```python
env = QProcessEnvironment.systemEnvironment()
```

Based on my reading of this, though, I think it's a copy anyway:

https://doc.qt.io/qtforpython/PySide6/QtCore/QProcessEnvironment.html#PySide6.QtCore.PySide6.QtCore.QProcessEnvironment.systemEnvironment

> This function does not cache the system environment. [...] note that repeated calls to this function will recreate the [QProcessEnvironment](https://doc.qt.io/qtforpython/PySide6/QtCore/QProcessEnvironment.html#PySide6.QtCore.PySide6.QtCore.QProcessEnvironment) object, which is a non-trivial operation.

Comment (psobolewskiPhD) [reply napari/_qt/dialogs/qt_package_installer.py]: Your test code fails because `TypeError: argument of type 'QProcessEnvironment' is not iterable`
However:
```
from qtpy.QtCore import QProcess, QProcessEnvironment

env = QProcessEnvironment.systemEnvironment()

env.value('PYTHONEXECUTABLE')
Out[3]: '/Users/piotrsobolewski/Dev/miniforge3/envs/napari-main/bin/python3.10'

env.remove('PYTHONEXECUTABLE')

env2 = QProcessEnvironment.systemEnvironment()

env2.value('PYTHONEXECUTABLE')
Out[7]: '/Users/piotrsobolewski/Dev/miniforge3/envs/napari-main/bin/python3.10'
```
So it seems fine.
I'm not sure what the suggestion is, something missing? Screenshot:
image

Comment (aganders3) [reply napari/_qt/dialogs/qt_package_installer.py]: I found another spot in the docs where it's very clear:

> The `QProcessEnvironment` class wraps that concept and allows easy manipulation of those variables. It’s meant to be used along with `QProcess` , to set the environment for child processes. It cannot be used to change the current process’s environment.

Comment (psobolewskiPhD) [reply napari/_qt/dialogs/qt_package_installer.py]: Cool thanks!
So what do y'all think of putting this behind a set of `if` like the existing code here:
https://github.com/napari/napari/blob/9e911040148d232b76d768ed74ff90ac0fa6f707/napari/_qt/dialogs/qt_package_installer.py#L207-L215

Comment (aganders3) [reply napari/_qt/dialogs/qt_package_installer.py]: Hm yeah. I think PYTHONEXECUTABLE only affects macOS anyway but I think putting it in a block down with the other platform-specific code (i.e. that "nt" block) would make the motivation more clear and self-documenting.

Edit: or do you think it needs to stay within this `if env is None`?

Comment (psobolewskiPhD) [reply napari/_qt/dialogs/qt_package_installer.py]: No, putting it in `if env is None` was just mindlessness and inexperience on my part.

I'll make a dedicated `if darwin` block for it.

**Comment by Czaki** [comment]
What happens in a scenario when a user has conda in a path, but the environment is not created with conda (ex. virtualenv)? Did the option to install from the conda is still exposed?

**Comment by psobolewskiPhD** [comment]
> What happens in a scenario when a user has conda in a path, but the environment is not created with conda (ex. virtualenv)? Did the option to install from the conda is still exposed?

I assume so? the logic is in the other file:
https://github.com/napari/napari/blob/9e911040148d232b76d768ed74ff90ac0fa6f707/napari/_qt/dialogs/qt_plugin_dialog.py#L560-L564

Looks like it's based on the widget which is determined by whether the plugin is on conda or not. 🤷‍♂️

**Comment by Czaki** [comment]
> Looks like it's based on the widget which is determined by whether the plugin is on conda or not. man_shrugging

This is installation logic, not detection if widget is even visible.

**Comment by psobolewskiPhD** [comment]
> This is installation logic, not detection if widget is even visible.

I can't figure that part out, sorry.

Edit: Ok, it's quite hard to follow the logic for me—it took 2 hours 😭—but I think it's actually this line:
https://github.com/napari/napari/blob/9e911040148d232b76d768ed74ff90ac0fa6f707/napari/_qt/dialogs/qt_plugin_dialog.py#L846

That calls a function in npe2api:
https://github.com/napari/napari/blob/9e911040148d232b76d768ed74ff90ac0fa6f707/napari/plugins/npe2api.py#L74-L101

The yield then goes here:
https://github.com/napari/napari/blob/9e911040148d232b76d768ed74ff90ac0fa6f707/napari/_qt/dialogs/qt_plugin_dialog.py#L1057-L1066

So then working back from that, we see the data is used here:
https://github.com/napari/napari/blob/9e911040148d232b76d768ed74ff90ac0fa6f707/napari/_qt/dialogs/qt_plugin_dialog.py#L1028-L1055

And there we see the call to `add_item` which is here:
https://github.com/napari/napari/blob/9e911040148d232b76d768ed74ff90ac0fa6f707/napari/_qt/dialogs/qt_plugin_dialog.py#L455-L495

And that's where the widget is made. As far as I can tell there is no checks for when the widget shows conda. Unless I've missed something, I'd expect the widget to show conda regardless of how the env was created.
If conda isn't actually available then I think the only checks are here:
https://github.com/napari/napari/blob/9e911040148d232b76d768ed74ff90ac0fa6f707/napari/_qt/dialogs/qt_package_installer.py#L166-L186

**Comment by Czaki** [comment]
> And that's where the widget is made. As far as I can tell there is no checks for when the widget shows conda. Unless I've missed something, I'd expect the widget to show conda regardless of how the env was created.
> If conda isn't actually available then I think the only checks are here:

So it looks like there is a bug in the logic. Maybe out of the scope of this PR but at least should have issue about this.

**Review by goanpeca — APPROVED** [review]

**Comment by psobolewskiPhD** [comment]
> > And that's where the widget is made. As far as I can tell there is no checks for when the widget shows conda. Unless I've missed something, I'd expect the widget to show conda regardless of how the env was created.
> > If conda isn't actually available then I think the only checks are here:
> 
> So it looks like there is a bug in the logic. Maybe out of the scope of this PR but at least should have issue about this.

I think for sure out of scope, this is a small bug fix for a rather niche case.
If indeed conda is offered when not available or offered when non-condo env is used, that's a bigger deal I think. But best someone without conda test this first to make sure?

**Review by jni — APPROVED** [review]

**Comment by psobolewskiPhD** [comment]
@czaki are you ok with this, given https://github.com/napari/napari/pull/5641
(you commented above, so just making sure)

**Comment by Czaki** [comment]
> @Czaki are you ok with this, given #5641
> (you commented above, so just making sure)

I'm ok. I do not have enough knowledge to validate it in mind and do not have a Mac to check it on the machine.
